File: src/core/theme_loader.zig
const std = @import("std");
const tui = @import("tui");
const Color = tui.Color;
const Theme = @import("theme.zig").Theme;
const hard_theme = @import("../themes/hard.zig").theme;

fn getStringField(obj: std.json.ObjectMap, key: []const u8) ?[]const u8 {
    const field = obj.get(key) orelse return null;
    if (field != .string) return null;
    return field.string;
}

/// Parse a hex color string like "#1b1b1b" into a Color.
/// Returns null for "default" or invalid strings.
fn parseColorField(value: ?[]const u8, default: Color) Color {
    const s = value orelse return default;
    if (std.mem.eql(u8, s, "default")) return .default;
    if (s.len != 7 or s[0] != '#') return default;
    const hex_val = std.fmt.parseInt(u24, s[1..], 16) catch return default;
    return Color.hex(hex_val);
}

fn loadFromJsonValue(value: std.json.Value) !Theme {
    if (value != .object) return error.InvalidThemeFormat;
    const obj = value.object;

    return Theme{
        .background = parseColorField(getStringField(obj, "background"), hard_theme.background),
        .foreground = parseColorField(getStringField(obj, "foreground"), hard_theme.foreground),
        .surface = parseColorField(getStringField(obj, "surface"), hard_theme.surface),
        .surface_alt = parseColorField(getStringField(obj, "surface_alt"), hard_theme.surface_alt),
        .primary = parseColorField(getStringField(obj, "primary"), hard_theme.primary),
        .primary_bright = parseColorField(getStringField(obj, "primary_bright"), hard_theme.primary_bright),
        .on_primary = parseColorField(getStringField(obj, "on_primary"), hard_theme.on_primary),
        .on_primary_bright = parseColorField(getStringField(obj, "on_primary_bright"), hard_theme.on_primary_bright),
        .success = parseColorField(getStringField(obj, "success"), hard_theme.success),
        .success_bright = parseColorField(getStringField(obj, "success_bright"), hard_theme.success_bright),
        .danger = parseColorField(getStringField(obj, "danger"), hard_theme.danger),
        .danger_bright = parseColorField(getStringField(obj, "danger_bright"), hard_theme.danger_bright),
        .warning = parseColorField(getStringField(obj, "warning"), hard_theme.warning),
        .info = parseColorField(getStringField(obj, "info"), hard_theme.info),
        .info_bright = parseColorField(getStringField(obj, "info_bright"), hard_theme.info_bright),
        .accent = parseColorField(getStringField(obj, "accent"), hard_theme.accent),
        .accent_bright = parseColorField(getStringField(obj, "accent_bright"), hard_theme.accent_bright),
        .emphasis = parseColorField(getStringField(obj, "emphasis"), hard_theme.emphasis),
        .muted = parseColorField(getStringField(obj, "muted"), hard_theme.muted),
        .forge_github = parseColorField(getStringField(obj, "forge_github"), hard_theme.forge_github),
        .forge_gitlab = parseColorField(getStringField(obj, "forge_gitlab"), hard_theme.forge_gitlab),
        .forge_codeberg = parseColorField(getStringField(obj, "forge_codeberg"), hard_theme.forge_codeberg),
    };
}

pub fn loadFromFile(allocator: std.mem.Allocator, io: std.Io, path: []const u8) !Theme {
    var file_buf: [8192]u8 = undefined;
    const content = std.Io.Dir.readFile(.cwd(), io, path, &file_buf) catch return error.ThemeReadError;

    var arena = std.heap.ArenaAllocator.init(allocator);
    defer arena.deinit();
    const arena_alloc = arena.allocator();

    const parsed = try std.json.parseFromSlice(std.json.Value, arena_alloc, content, .{});
    defer parsed.deinit();

    return try loadFromJsonValue(parsed.value);
}

pub fn customThemePath(allocator: std.mem.Allocator, environ_map: *std.process.Environ.Map, name: []const u8) !?[]u8 {
    if (!isSafeCustomThemeName(name)) return null;
    const home = environ_map.get("HOME") orelse return null;
    return try std.fmt.allocPrint(allocator, "{s}/.config/flavors-tmux/themes/{s}.json", .{ home, name });
}

pub fn customThemePathBuf(environ_map: *std.process.Environ.Map, name: []const u8, buf: []u8) !?[]const u8 {
    if (!isSafeCustomThemeName(name)) return null;
    const home = environ_map.get("HOME") orelse return null;
    return try std.fmt.bufPrint(buf, "{s}/.config/flavors-tmux/themes/{s}.json", .{ home, name });
}

pub fn isSafeCustomThemeName(name: []const u8) bool {
    if (name.len == 0) return false;
    for (name) |c| {
        if (std.ascii.isAlphanumeric(c) or c == '_' or c == '-') continue;
        return false;
    }
    return true;
}

test "custom theme names reject path traversal" {
    try std.testing.expect(isSafeCustomThemeName("my_theme-1"));
    try std.testing.expect(!isSafeCustomThemeName("../hard"));
    try std.testing.expect(!isSafeCustomThemeName("nested/theme"));
    try std.testing.expect(!isSafeCustomThemeName(""));
}

test "custom themes fall back to hard colors for missing keys" {
    var arena = std.heap.ArenaAllocator.init(std.testing.allocator);
    defer arena.deinit();

    const parsed = try std.json.parseFromSlice(std.json.Value, arena.allocator(), "{\"background\":\"#123456\"}", .{});
    defer parsed.deinit();

    const theme = try loadFromJsonValue(parsed.value);
    try std.testing.expect(std.meta.eql(Color.hex(0x123456), theme.background));
    try std.testing.expect(std.meta.eql(hard_theme.foreground, theme.foreground));
    try std.testing.expect(std.meta.eql(hard_theme.surface_alt, theme.surface_alt));
}

File: src/core/theme_loader.rs
use std::collections::HashMap;
use std::env;
use std::fs;
use std::path::{Path, PathBuf};
use std::sync::Mutex;
use std::time::SystemTime;

use serde::Deserialize;

use crate::core::{Color, Theme};
use crate::themes::hard;

/// In-memory cache of custom themes keyed by file path + mtime.
/// Avoids repeated file I/O when multiple widgets resolve the same
/// custom theme within a single status refresh.
static CUSTOM_THEME_CACHE: Mutex<Option<HashMap<PathBuf, (SystemTime, Theme)>>> = Mutex::new(None);

#[derive(Debug)]
pub enum ThemeLoaderError {
    Read(std::io::Error),
    Json(serde_json::Error),
    #[allow(dead_code)]
    InvalidThemeFormat,
}

impl From<std::io::Error> for ThemeLoaderError {
    fn from(value: std::io::Error) -> Self {
        Self::Read(value)
    }
}

impl From<serde_json::Error> for ThemeLoaderError {
    fn from(value: serde_json::Error) -> Self {
        Self::Json(value)
    }
}

#[derive(Deserialize)]
struct PartialTheme {
    background: Option<String>,
    foreground: Option<String>,
    surface: Option<String>,
    surface_alt: Option<String>,
    primary: Option<String>,
    primary_bright: Option<String>,
    on_primary: Option<String>,
    on_primary_bright: Option<String>,
    success: Option<String>,
    success_bright: Option<String>,
    danger: Option<String>,
    danger_bright: Option<String>,
    warning: Option<String>,
    info: Option<String>,
    info_bright: Option<String>,
    accent: Option<String>,
    accent_bright: Option<String>,
    emphasis: Option<String>,
    muted: Option<String>,
    forge_github: Option<String>,
    forge_gitlab: Option<String>,
    forge_codeberg: Option<String>,
}

fn parse_color_field(value: Option<&str>, default: Color) -> Color {
    let Some(raw) = value else {
        return default;
    };
    if raw == "default" {
        return Color::Default;
    }
    let Some(hex) = raw.strip_prefix('#') else {
        return default;
    };
    if hex.len() != 6 {
        return default;
    }
    match u32::from_str_radix(hex, 16) {
        Ok(value) => Color::hex(value),
        Err(_) => default,
    }
}

fn load_from_partial(partial: &PartialTheme) -> Theme {
    let default = hard::THEME;
    Theme {
        background: parse_color_field(partial.background.as_deref(), default.background),
        foreground: parse_color_field(partial.foreground.as_deref(), default.foreground),
        surface: parse_color_field(partial.surface.as_deref(), default.surface),
        surface_alt: parse_color_field(partial.surface_alt.as_deref(), default.surface_alt),
        primary: parse_color_field(partial.primary.as_deref(), default.primary),
        primary_bright: parse_color_field(
            partial.primary_bright.as_deref(),
            default.primary_bright,
        ),
        on_primary: parse_color_field(partial.on_primary.as_deref(), default.on_primary),
        on_primary_bright: parse_color_field(
            partial.on_primary_bright.as_deref(),
            default.on_primary_bright,
        ),
        success: parse_color_field(partial.success.as_deref(), default.success),
        success_bright: parse_color_field(
            partial.success_bright.as_deref(),
            default.success_bright,
        ),
        danger: parse_color_field(partial.danger.as_deref(), default.danger),
        danger_bright: parse_color_field(partial.danger_bright.as_deref(), default.danger_bright),
        warning: parse_color_field(partial.warning.as_deref(), default.warning),
        info: parse_color_field(partial.info.as_deref(), default.info),
        info_bright: parse_color_field(partial.info_bright.as_deref(), default.info_bright),
        accent: parse_color_field(partial.accent.as_deref(), default.accent),
        accent_bright: parse_color_field(partial.accent_bright.as_deref(), default.accent_bright),
        emphasis: parse_color_field(partial.emphasis.as_deref(), default.emphasis),
        muted: parse_color_field(partial.muted.as_deref(), default.muted),
        forge_github: parse_color_field(partial.forge_github.as_deref(), default.forge_github),
        forge_gitlab: parse_color_field(partial.forge_gitlab.as_deref(), default.forge_gitlab),
        forge_codeberg: parse_color_field(
            partial.forge_codeberg.as_deref(),
            default.forge_codeberg,
        ),
    }
}

pub fn load_from_json_str(content: &str) -> Result<Theme, ThemeLoaderError> {
    let partial: PartialTheme = serde_json::from_str(content)?;
    Ok(load_from_partial(&partial))
}

pub fn load_from_file(path: &Path) -> Result<Theme, ThemeLoaderError> {
    let mtime = path.metadata()?.modified()?;
    if let Ok(cache) = CUSTOM_THEME_CACHE.lock() {
        if let Some(map) = cache.as_ref() {
            if let Some((cached_mtime, cached_theme)) = map.get(path) {
                if *cached_mtime == mtime {
                    return Ok(*cached_theme);
                }
            }
        }
    }
    let content = fs::read_to_string(path)?;
    let theme = load_from_json_str(&content)?;
    if let Ok(mut cache) = CUSTOM_THEME_CACHE.lock() {
        let map = cache.get_or_insert_with(HashMap::new);
        if map.len() > 64 {
            map.clear();
        }
        map.insert(path.to_path_buf(), (mtime, theme));
    }
    Ok(theme)
}

pub fn custom_theme_path(name: &str) -> Option<PathBuf> {
    if !is_safe_custom_theme_name(name) {
        return None;
    }
    let home = env::var_os("HOME")?;
    let mut path = PathBuf::from(home);
    path.push(".config");
    path.push("flavors-tmux");
    path.push("themes");
    path.push(format!("{name}.json"));
    Some(path)
}

pub fn is_safe_custom_theme_name(name: &str) -> bool {
    !name.is_empty()
        && name
            .bytes()
            .all(|byte| byte.is_ascii_alphanumeric() || byte == b'_' || byte == b'-')
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn custom_theme_names_reject_path_traversal() {
        assert!(is_safe_custom_theme_name("my_theme-1"));
        assert!(!is_safe_custom_theme_name("../hard"));
        assert!(!is_safe_custom_theme_name("nested/theme"));
        assert!(!is_safe_custom_theme_name(""));
    }

    #[test]
    fn custom_themes_fall_back_to_hard_colors_for_missing_keys() -> Result<(), ThemeLoaderError> {
        let theme = load_from_json_str("{\"background\":\"#123456\"}")?;
        assert_eq!(Color::hex(0x123456), theme.background);
        assert_eq!(hard::THEME.foreground, theme.foreground);
        assert_eq!(hard::THEME.surface_alt, theme.surface_alt);
        Ok(())
    }
}
